app/models/events/tags.rb

```diff
@@ -0,0 +1,39 @@
+module Events::Tags
+  extend ActiveSupport::Concern
+
+  included do
+    validates :tag, presence: true, inclusion: { in: ->(_) { tags.keys } }
+    attribute :tag, Events::Tag::Type.new
+    validates :additional, nested: true, allow_nil: true
+    belongs_to :ip_address, optional: true
+    belongs_to :geoip_info, optional: true
+
+    cattr_reader(:tags) { {} }
+
+    def additional_type
+      tags.fetch(tag, nil)
+    end
+
+    def additional
+      additional_type&.new(super) || super
+    end
+
+    def additional=(value)
+      super(value&.to_h)
+    end
+  end
+
+  class_methods do
+    def define_event(string, &blk)
+      tag = Events::Tag.new(string).freeze
+      raise ArgumentError, "Tag #{tag.inspect} already defined" if tags.key?(tag)
+      event = Events::Tag.to_struct(&blk)
+      tags[tag] = event
```

Comment:
Could we shift to using the string as key? It's is actually the key since that's how we store and retrieve it from the database. Then deserializing might just be "look it up in the hash table" instead of reconstructing.

Reply:
except we use the methods on it elsewhere, and then we have to write a custom active record serializer to handle the lookup into the map

Reply:
Unless I misunderstood, the methods we use are string manipulations of the key. The code would not change much if they were class methods. It would certainly save complexity and remove a lot of low level code.

I could be totally wrong, you spent more time in this code than me. Just noting complexity where I see it.